### PR TITLE
MODREQMED-48: Mediated request confirmation - create ECS TLR on behalf of fake user

### PR DESCRIPTION
**Pull request by imerabishvili**
## Purpose
https://folio-org.atlassian.net/browse/MODREQMED-48

## Approach
Before creating an ECS TLR for non-local instance/item case:
1. Create a fake user
2. Create a DB record to link real requester and fake user created in (1)
3. Pass fakeUser.id as a requester when craeting ECS TLR

### CODE REVIEW

File: src/main/java/org/folio/mr/service/impl/MediatedRequestActionsServiceImpl.java [hunk 1]
```diff
@@ -69,22 +68,35 @@ private void createRequest(MediatedRequestEntity mediatedRequest) {
...
       EcsTlr ecsTlr = ecsRequestService.create(mediatedRequest);
-      mediatedRequest.setConfirmedRequestId(UUID.fromString(ecsTlr.getPrimaryRequestId()));
+      Request request = updatePrimaryRequest(mediatedRequest, ecsTlr);
+      updateMediatedRequest(mediatedRequest, ecsTlr, request);
     }
   }
 
+  private Request updatePrimaryRequest(MediatedRequestEntity mediatedRequest, EcsTlr ecsTlr) {
+    Request request = circulationRequestService.get(ecsTlr.getPrimaryRequestId());
+    request.setRequesterId(mediatedRequest.getRequesterId().toString());
```
Comment (roman-barannyk) [inline src/main/java/org/folio/mr/service/impl/MediatedRequestActionsServiceImpl.java:78]: can it be null?

File: src/main/resources/db/changelog/changelog-1.0.0.xml [hunk 2]
```diff
@@ -10,4 +10,5 @@
     <include file="changes/v1.0.0/2024-08-06-add-awaiting-delivery-step.xml" relativeToChangelogFile="true"/>
     <include file="changes/v1.0.0/2024-08-30-create-mediated_request_workflow_log-table.xml" relativeToChangelogFile="true"/>
     <include file="changes/v1.0.0/2024-09-18-make-some-mediated_request-columns-nullable.xml" relativeToChangelogFile="true"/>
-</databaseChangeLog>
\ No newline at end of file
+    <include file="changes/v1.0.0/2024-11-21-create_fake_patron_link-table.xml" relativeToChangelogFile="true"/>
```
Comment (roman-barannyk) [inline src/main/resources/db/changelog/changelog-1.0.0.xml:13]: I believe it should be v1.1.0 package

File: src/test/java/org/folio/mr/api/MediatedRequestActionsApiTest.java [hunk 3]
```diff
@@ -132,6 +135,7 @@ void mediatedRequestConfirmationForLocalInstanceAndItem() {
   @Test
   @SneakyThrows
   void mediatedRequestConfirmationForLocalInstanceAndRemoteItem() {
+    // given
```
Comment (roman-barannyk) [inline src/test/java/org/folio/mr/api/MediatedRequestActionsApiTest.java:138]: I think these messages do not make a lot of sense

File: src/main/java/org/folio/mr/client/CirculationClient.java [hunk 4]
```diff
@@ -3,13 +3,22 @@
...
   configuration = FeignClientConfiguration.class)
 public interface CirculationClient {
 
+  @GetMapping("/requests/{id}")
+  Request getRequest(@PathVariable String id);
+
   @PostMapping("/requests")
   Request createRequest(Request request);
 
+  @PutMapping("/requests")
```
[the remark this answers is not in the document]
Comment (imerabishvili) [reply src/main/java/org/folio/mr/client/CirculationClient.java]: fixed

File: src/test/java/org/folio/mr/api/MediatedRequestActionsApiTest.java (hunk 3 above)
Comment (imerabishvili) [reply src/test/java/org/folio/mr/api/MediatedRequestActionsApiTest.java:138]: I found them useful in complex tests where I can't understand where preparation ends.
https://en.wikipedia.org/wiki/Given-When-Then

File: src/main/resources/db/changelog/changelog-1.0.0.xml (hunk 2 above)
Comment (imerabishvili) [reply src/main/resources/db/changelog/changelog-1.0.0.xml:13]: @roman-barannyk changes are applied to rancher DB so I can't move the files now.

File: src/main/java/org/folio/mr/service/impl/EcsRequestServiceImpl.java [hunk 5]
```diff
@@ -1,50 +1,103 @@
...
+      .active(true)
+      .barcode(format("securepatron_%s", UUID.randomUUID()))
+      .type("patron")
+      .patronGroup(patronGroupId)
+      .personal(
+        new UserPersonal()
+          .preferredContactTypeId("002")
+          .firstName("Secure")
+          .lastName("Patron")
+          .email(UUID.randomUUID() + "@example.com")
```
Comment (alexanderkurash) [inline src/main/java/org/folio/mr/service/impl/EcsRequestServiceImpl.java:75]: Could be the same as barcode, `@example.com` part is not necessary

File: src/main/java/org/folio/mr/service/impl/CirculationRequestServiceImpl.java [hunk 6]
```diff
@@ -39,4 +45,10 @@ private static Request buildRequest(MediatedRequestEntity mediatedRequest) {
       .deliveryAddressTypeId(asString(mediatedRequest.getDeliveryAddressTypeId()))
       .patronComments(mediatedRequest.getPatronComments());
   }
+
+  @Override
+  public Request update(Request request) {
+    log.info("update:: update circulation request {}", request);
```
[the remark this answers is not in the document]
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/CirculationRequestServiceImpl.java]: fixed

File: src/main/java/org/folio/mr/service/impl/EcsRequestServiceImpl.java (hunk 5 above)
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/EcsRequestServiceImpl.java]: done

File: src/main/java/org/folio/mr/domain/entity/FakePatronLink.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
+package org.folio.mr.domain.entity;
+
+import jakarta.persistence.Entity;
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/mr/domain/entity/FakePatronLink.java:3]: We use different import order in our projects:
static imports
java.*
javax.*
org.*
com.*
all other imports

File: src/main/java/org/folio/mr/service/impl/MediatedRequestActionsServiceImpl.java (hunk 1 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/mr/service/impl/MediatedRequestActionsServiceImpl.java:78]: No, 
<column name="requester_id" type="uuid">
                <constraints nullable="false"/>
            </column>

File: src/test/java/org/folio/mr/api/MediatedRequestActionsApiTest.java [hunk 8]
```diff
@@ -154,12 +158,30 @@ void mediatedRequestConfirmationForLocalInstanceAndRemoteItem() {
       .withHeader(HEADER_TENANT, equalTo(TENANT_ID_CENTRAL))
       .willReturn(jsonResponse(new ConsortiumItems().items(new ArrayList<>()), HttpStatus.SC_OK)));
 
+    User user = new User().id(UUID.randomUUID().toString());
+    wireMockServer.stubFor(WireMock.post(urlMatching(USERS_URL_TEMPLATE))
+      .withHeader(HEADER_TENANT, equalTo(TENANT_ID_CENTRAL))
+      .willReturn(jsonResponse(user, HttpStatus.SC_OK)));
+
+    wireMockServer.stubFor(WireMock.post(urlMatching(USERS_URL_TEMPLATE))
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/mr/api/MediatedRequestActionsApiTest.java:166]: why do we need a second stub for the same URL?

File: src/test/java/org/folio/mr/service/UserCloningServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+  @Mock
+  private UserService userService;
+
+  @InjectMocks
+  private UserCloningServiceImpl userCloningService;
+
+  @Test
+  void testCloneWhenUserExists() {
+    // given
```
Comment (OleksandrVidinieiev) [inline src/test/java/org/folio/mr/service/UserCloningServiceTest.java:29]: Do we really need these comments in a 4-line test?

File: src/main/java/org/folio/mr/service/impl/EcsRequestServiceImpl.java [hunk 10]
```diff
@@ -1,50 +1,103 @@
...
+  private String createFakePatron(MediatedRequestEntity mediatedRequest) {
+    User realPatron = userService.fetchUser(asString(mediatedRequest.getRequesterId()));
+    User fakePatron = buildFakePatron(realPatron.getPatronGroup());
+
+    log.info("createFakePatron:: Creating local fake patron");
+    User localFake = userService.create(fakePatron);
+
+    log.info("createFakePatron:: Creating remote fake shadow patron: ID {}", localFake.getId());
+    executionService.executeSystemUserScoped(consortiumService.getCentralTenantId(),
+      () -> userCloningService.cloneUser(localFake, consortiumService.getCurrentTenantId()));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/mr/service/impl/EcsRequestServiceImpl.java:58]: No need to clone the requester, mod-tlr will do it for you

File: src/test/java/org/folio/mr/api/MediatedRequestActionsApiTest.java (hunk 8 above)
Comment (imerabishvili) [reply src/test/java/org/folio/mr/api/MediatedRequestActionsApiTest.java:166]: Yes user is created twice - one for current tenant and one in central one.

File: src/test/java/org/folio/mr/service/UserCloningServiceTest.java (hunk 9 above)
Comment (imerabishvili) [reply src/test/java/org/folio/mr/service/UserCloningServiceTest.java:29]: removed

File: src/main/java/org/folio/mr/service/impl/EcsRequestServiceImpl.java (hunk 10 above)
Comment (imerabishvili) [reply src/main/java/org/folio/mr/service/impl/EcsRequestServiceImpl.java]: Requester is not cloned

**Review by alexanderkurash — APPROVED** [review]

**Review by roman-barannyk — APPROVED** [review]

**Pull request merged**